File: i3/.i3/connection_toggle.py
#!/usr/bin/env python

import sh

# get table of connections from nmcli
connections = str(sh.nmcli(
       "--colors", "no",
       "--fields", "name,type,state",
       "connection", "show"))

# split the table into header and rows
header, rows = connections.split("\n", 1)

# use rofi promt to chose connection
selection = sh.rofi(sh.echo(rows), "-dmenu",
       "-i",
       "-p", "Toggle Connection",
       "-mesg", header)

name = selection.split()[0]
state = selection.split()[-1]

# toggle connection
command = "up" if state == "--" else "down"
sh.nmcli("connection", command, name)

File: pkgs/connection-toggle.clj
(ns connection-toggle
  (:require [clojure.java.shell :refer [sh]]
            [clojure.string :as str]))

;; get table of connections from nmcli
(def table (-> (sh "nmcli"
                   "--colors" "no"
                   "--fields" "name,type,state"
                   "connection" "show")
               :out
               str/split-lines))

;; use rofi promt to chose connection
(def selection (let [header (first table)
                     rows (rest table)
                     {:keys [out exit]}
                     (sh "rofi" "-dmenu"
                         "-i"                      ;; case insensitive
                         "-p" "Toggle Connection"  ;; prompt
                         "-mesg" header
                         :in (str/join "\n" rows))]
                 (when (zero? exit) out)))

;; toggle connection
(let [[name _ state] (some-> selection (str/split  #"\s{2,}"))
      action (if (= "--" state) "up" "down")]
  (when (some? state)
    (sh "nmcli" "connection" action name)))


